src/browser-extension/popup/BrowserPopup.tsx

```diff
@@ -1,41 +1,180 @@
-import React from 'react';
+import React, { ChangeEvent } from 'react';
+
+import addIcon from '../../../assets/pp-add-icon.svg';
+import requestIcon from '../../../assets/pp-request-icon.svg';
+import switchOffIcon from '../../../assets/pp-switch-off-icon.svg';
+
+import { standardizeUrlForPageSettings } from 'utils/url';
+import Toggle from './toggle/toggle';
+import chromeStorage from 'chrome-storage';
+import * as chromeKeys from 'chrome-storage/keys';
+import _filter from 'lodash/filter';
+import classNames from 'classnames';
 
 declare const chrome: any;
 
-export default class BrowserPopup extends React.Component<{}, {}> {
+interface IBrowserPopupState {
+  isLoading: boolean;
+  currentTabUrl: string;
+  annotationModePages: string[];
+  isExtensionDisabled: boolean;
+  disabledPages: string[];
+}
+
+export default class BrowserPopup extends React.Component<{}, Partial<IBrowserPopupState>> {
   constructor(props: {}) {
     super(props);
+
+    this.state = {
+      isLoading: true,
+      currentTabUrl: null,
+      annotationModePages: [],
+      isExtensionDisabled: false,
+      disabledPages: [],
+    };
+  }
+
+  loadStateFromStorage() {
+    /* Use data from chrome storage to update popup state
+     * It is assumed that there are no other sources of data;
+     * In the future:
+     *   for more sources of data, resort to a local popup Redux store or a global background page Redux store
+     */
+    if (this.state.isLoading) {
+      chromeStorage.get([
+        chromeKeys.ANNOTATION_MODE_PAGES,
+        chromeKeys.DISABLED_EXTENSION,
+        chromeKeys.DISABLED_PAGES,
+      ], (result) => {
+        this.setState({
+          isLoading: false,
+          annotationModePages: result[chromeKeys.ANNOTATION_MODE_PAGES] || [],
+          isExtensionDisabled: result[chromeKeys.DISABLED_EXTENSION] || false,
+          disabledPages: result[chromeKeys.DISABLED_PAGES] || [],
+        });
+      });
+    }
+
+  }
+
+  componentDidMount() {
+    chrome.tabs.query({ active: true, currentWindow: true }, (tabs) => {
+      const tab = tabs[0];
+      this.setState({ currentTabUrl: standardizeUrlForPageSettings(tab.url) });
+    });
+
+    this.loadStateFromStorage();
+  }
+
+  componentDidUpdate() {
+    this.loadStateFromStorage();
+  }
+
+  isExtensionDisabledForCurrentTab() {
+    return (this.state.disabledPages || []).indexOf(this.state.currentTabUrl) !== -1;
```

Comment:
I think using [includes](https://devdocs.io/javascript/global_objects/array/includes) (as it is finally available) is a bit more readable

Reply:
Thanks, didn't know about it :+1: